src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+import * as path from "path";
+import * as vscode from "vscode";
+import {
+    DocumentSelector,
+    LanguageClient,
+} from "vscode-languageclient";
+import { IFeature } from "../feature";
+import * as Settings from "../settings";
+
+const tm = getCoreNodeModule("vscode-textmate");
+// Returns a node module installed with VSCode, or null if it fails.
+// https://github.com/Microsoft/vscode/issues/46281
+function getCoreNodeModule(moduleName: string) {
+    try {
+        return require(`${vscode.env.appRoot}/node_modules.asar/${moduleName}`);
+    // tslint:disable-next-line:no-empty
+    } catch (err) { }
+
+    try {
+        return require(`${vscode.env.appRoot}/node_modules/${moduleName}`);
+    // tslint:disable-next-line:no-empty
+    } catch (err) { }
+
+    return null;
+}
+
+interface IExtensionGrammar {
+    language?: string;
+    scopeName?: string;
+    path?: string;
+    extensionPath?: string;
+    embeddedLanguages?: {
+        [scopeName: string]: string,
+    }; injectTo?: string[];
+}
+
+interface IExtensionLanguage {
+    id: string;
+    configuration: string;
+}
+
+interface IExtensionPackage {
+    contributes?: {
+        languages?: IExtensionLanguage[],
+        grammars?: IExtensionGrammar[],
+    };
+}
+
+// Need to reproduce the IToken interface from vscode-textmate due to
+// the odd way it has to be required
+// https://github.com/Microsoft/vscode-textmate/blob/46af9487e1c8fa78aa1f2e2/release/main.d.ts#L161-L165
+export interface IToken {
+    startIndex: number;
+    readonly endIndex: number;
+    readonly scopes: string[];
+}
+
+// The different types of matched textmate tokens.
+// Mirrors the enum at https://code.visualstudio.com/docs/extensionAPI/vscode-api#FoldingRangeKind
+enum MatchType {
+    Comment = 1,
+    Imports = 2,
+    Region = 3,
+}
+
+// Defines a pair of tokens (start, end) or pair or line numbers (startLine, endLine) which define
+// a folding range in source code.
+class MatchedToken {
+    // Start IToken object of the range.  Optional if startLine is specified
+    public start: IToken;
+    // The starting zero-based line number of the range.
+    public startline: number;
+    // End IToken object of the range.  Optional if startLine is specified
+    public end: IToken;
+    // The ending zero-based line number of the range.
+    public endline: number;
+    // The type of range this is
+    public matchType: MatchType;
+
+    constructor(start: IToken = null,
+                end: IToken = null,
+                startLine: number = null,
+                endLine: number = null,
+                matchType: MatchType,
+                document: vscode.TextDocument) {
+        this.start = start;
+        this.end = end;
+        if (startLine === null) {
```

Comment:
I was going to suggest `||`, but both suggestions are wrong since `startLine` can be `0`.

Needs to be:
```typescript
this.startLine = startLine !== null ? startLine : document.positionAt(start.startIndex).line;
```

And if there's a worry about the possibility of `undefined` being a possible value of `startLine`, `==` will work for that.